src/ledger/LedgerState.cpp

```diff
@@ -0,0 +1,801 @@
+// Copyright 2017 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "ledger/LedgerState.h"
+#include "crypto/Hex.h"
+#include "crypto/SHA.h"
+#include "ledger/LedgerEntryReference.h"
+#include "ledger/LedgerHeaderReference.h"
+#include "xdr/Stellar-ledger.h"
+#include "xdrpp/marshal.h"
+
+// TODO(jonjove): Remove this header (used for LedgerEntryKey)
+#include "ledger/EntryFrame.h"
+
+namespace stellar
+{
+using xdr::operator==;
+
+bool
+operator==(InflationVotes const& lhs, InflationVotes const& rhs)
+{
+    return (lhs.votes == rhs.votes) && (lhs.inflationDest == rhs.inflationDest);
+}
+
+LedgerStateRoot::LedgerStateRoot(Database& db) : mHasChild(false), mDb(db)
+{
+}
+
+void
+LedgerStateRoot::hasChild(bool child)
+{
+    assert(!child || !mHasChild);
+    mHasChild = child;
+}
+
+LedgerHeader const&
+LedgerStateRoot::getPreviousHeader()
+{
+    return mPreviousHeader;
+}
+
+void
+LedgerStateRoot::setPreviousHeader(LedgerHeader const& header)
+{
+    mPreviousHeader = header;
+}
+
+Database&
+LedgerStateRoot::getDatabase()
+{
+    return mDb;
+}
+
+Database::EntryCache&
+LedgerStateRoot::getCache()
+{
+    // TODO(jonjove): Do we want to continue using the cache from Database?
+    // We could refactor so the cache is a member of LedgerStateRoot
+    return mDb.getEntryCache();
+}
+
+LedgerState::LedgerState(LedgerStateRoot& root)
+    : mRoot(&root), mParent(nullptr), mChild(nullptr)
+{
+    mRoot->hasChild(true);
+}
+
+LedgerState::LedgerState(LedgerState& parent)
+    : mRoot(nullptr), mParent(&parent), mChild(nullptr)
+{
+    assert(!mParent->mChild);
+    mParent->mChild = this;
+    for (auto const& state : mParent->mState)
+    {
+        state.second->invalidate();
+    }
+    mParent->mLoadBestOfferContext.clear();
+}
+
+LedgerState::~LedgerState()
+{
+    if (mParent || mRoot)
```

Comment:
is it possible to be true? I see only 2 constructors here and one always sets mParant and second one mRoot
maybe assert would be better?

Reply:
This comes into play if `LedgerState::rollback` or `LedgerState::commit` has already been called. When either of those occur, we effectively detach the nested `LedgerState` from the rest of the chain. In this case, `mParent == mRoot == nullptr`. This prevents `LedgerState` from rolling back if it has already been handled. An analogous assert appears in `rollback` and `commit`.